### PR TITLE
cmd/govim: refactor file watching

### LINKED ISSUE #492: cmd/govim: race condition in file watcher
### What version of Vim/Gvim are you using?

<details><summary>Vim/Gvim version</summary><br><pre>
$ vim --version
VIM - Vi IMproved 8.1 (2018 May 18, compiled Jul 22 2019 00:43:56)
macOS version
Included patches: 1-1729
Compiled by pontus@MacBook-Pro.localdomain
Huge version without GUI.  Features included (+) or not (-):
+acl               -farsi             -mouse_sysmouse    -tag_any_white
+arabic            +file_in_path      +mouse_urxvt       -tcl
+autocmd           +find_in_path      +mouse_xterm       +termguicolors
+autochdir         +float             +multi_byte        +terminal
-autoservername    +folding           +multi_lang        +terminfo
-balloon_eval      -footer            -mzscheme          +termresponse
+balloon_eval_term +fork()            +netbeans_intg     +textobjects
-browse            -gettext           +num64             +textprop
++builtin_terms    -hangul_input      +packages          +timers
+byte_offset       +iconv             +path_extra        +title
+channel           +insert_expand     -perl              -toolbar
+cindent           +job               +persistent_undo   +user_commands
-clientserver      +jumplist          +postscript        +vartabs
+clipboard         +keymap            +printer           +vertsplit
+cmdline_compl     +lambda            +profile           +virtualedit
+cmdline_hist      +langmap           -python            +visual
+cmdline_info      +libcall           -python3           +visualextra
+comments          +linebreak         +quickfix          +viminfo
+conceal           +lispindent        +reltime           +vreplace
+cryptv            +listcmds          +rightleft         +wildignore
-cscope            +localmap          -ruby              +wildmenu
+cursorbind        -lua               +scrollbind        +windows
+cursorshape       +menu              +signs             +writebackup
+dialog_con        +mksession         +smartindent       -X11
+diff              +modify_fname      -sound             -xfontset
+digraphs          +mouse             +spell             -xim
-dnd               -mouseshape        +startuptime       -xpm
-ebcdic            +mouse_dec         +statusline        -xsmp
+emacs_tags        -mouse_gpm         -sun_workshop      -xterm_clipboard
+eval              -mouse_jsbterm     +syntax            -xterm_save
+ex_extra          +mouse_netterm     +tag_binary        
+extra_search      +mouse_sgr         -tag_old_static    
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H   -DMACOS_X -DMACOS_X_DARWIN  -O2   -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L/usr/local/lib -o vim        -lm -lncurses  -liconv -framework AppKit        
</pre></details>

### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.13 darwin/amd64
</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
-
</pre></details>

### What version/commit of `govim` are you using?
956c1ebcc75

### What did you do?
Create a new package (dir) outside `vim` and directly copy a `.go` file into it.

### What did you expect to see?
File creation event for the new file.

### What did you see instead?
The current lib used for file watching doesn't support recursive directories. Instead new directories are captured and added to the watcher on the fly.

There is however a race condition since new files created in that directory between creation & the watcher being updated will be lost.

### PR DESCRIPTION
**Pull request by leitzler**
The previous file watchers was broken in several ways, where the two
major was:
* A race condition in the non-darwin watcher (#492)
* Darwin recursively watched the entire module directory, disregarding
  sub-modules, testdata directories and directories prefixed by _ or .
  This caused the test suite to always fail on macOS.

This change:
* Moves the non-recursive handling into os_other.go
* Adds an OS independent function that filter events specifically
  disallowed by Go (i.e. "testdata", sub-modules and directories
  prefixed by _ or .
* Adds more tests to raise file watching cover

Closes #492

### CODE REVIEW

File: cmd/govim/internal/fswatcher/os_darwin.go [hunk 1]
```diff
@@ -68,6 +71,10 @@ func New(gomodpath string, tomb *tomb.Tomb) (*FSWatcher, error) {
 				event := events[i]
 				path := filepath.Join(mountPoint, event.Path)
 
+				if skipDir(path) {
```
Comment (myitcv) [inline cmd/govim/internal/fswatcher/os_darwin.go:74]: Is this guaranteed to be a directory path we are passing here?

File: cmd/govim/watcher.go [hunk 2]
```diff
@@ -165,6 +87,39 @@ func (m *modWatcher) watch() {
...
 }
 
+func eventFilter(root string) func(string) bool {
+	return func(path string) bool {
+		if fi, err := os.Stat(path); err == nil && !fi.IsDir() {
+			path = filepath.Dir(path)
+		}
+
+		// TODO: cache ignored directories to reduce amount of os.Stat(...) calls
+		rel := strings.TrimPrefix(path, root)
```
Comment (myitcv) [inline cmd/govim/watcher.go:97]: I would personally do another `strings.TrimPrefix` after this line with `os.PathSeparator` to get you the relative path. Saves the rather obscure logic below of starting from 1. Also, it's my understanding (and I would guess yours too) that the path we will be passed is clean, correct?

File: cmd/govim/watcher.go [hunk 3]
```diff
@@ -165,6 +87,39 @@ func (m *modWatcher) watch() {
...
+		parts := strings.Split(rel, string(os.PathSeparator))
+		if len(parts) == 0 {
+			return false
+		}
+
+		curr := parts[0]
+		for i := 1; i < len(parts); i++ {
+			dir := parts[i]
+			switch dir[0] {
+			case '.', '_':
```
Comment (myitcv) [inline cmd/govim/watcher.go:107]: Can we link to the relevant part of the Go source code/docs for the reference on this list of conditions?

File: cmd/govim/watcher.go [hunk 4]
```diff
@@ -165,6 +87,39 @@ func (m *modWatcher) watch() {
 	}
 }
 
+func eventFilter(root string) func(string) bool {
+	return func(path string) bool {
+		if fi, err := os.Stat(path); err == nil && !fi.IsDir() {
```
Comment (myitcv) [inline cmd/govim/watcher.go:92]: What happens for _files_ that start with `.` or `_`? Are they notified?

**Review by myitcv — COMMENTED** [review]
I'll have a closer look later, but just a couple of initial questions.

File: cmd/govim/internal/fswatcher/os_darwin.go (hunk 1 above)
Comment (leitzler) [reply cmd/govim/internal/fswatcher/os_darwin.go:74]: No, it could be directories or files. The function name came from the fact that it did check the directory part of any path provided, but that's just an implementation detail. Maybe just calling it `filter(...)` is more clear.
```go
if filter(path) {
```

File: cmd/govim/watcher.go (hunk 2 above)
Comment (leitzler) [reply cmd/govim/watcher.go:97]: Interesting, that obscure logic looks like a left-over from something else. I think that the original intent was to avoid checking for the root `go.mod` file but that won't happen anyway. I'll update.

File: cmd/govim/watcher.go (hunk 3 above)
Comment (leitzler) [reply cmd/govim/watcher.go:107]: Good idea, will do.

File: cmd/govim/watcher.go (hunk 4 above)
Comment (leitzler) [reply cmd/govim/watcher.go:92]: Yes. I missed the part about files. Need to update so that the check applies to both directories and files.

File: cmd/govim/watcher.go (hunk 3 above)
Comment (leitzler) [reply cmd/govim/watcher.go:107]: As note to self - https://golang.org/cmd/go/#hdr-Package_lists_and_patterns

File: cmd/govim/internal/fswatcher/os_darwin.go (hunk 1 above)
Comment (myitcv) [reply cmd/govim/internal/fswatcher/os_darwin.go:74]: Agreed

**Comment by leitzler** [comment]
Updated per your comments, PTAL when you have time.

File: cmd/govim/internal/fswatcher/os_others.go [hunk 5]
```diff
@@ -13,49 +16,157 @@ type fswatcher struct {
...
+func New(root string, filter watchFilterFn, logf logFn, tomb *tomb.Tomb) (*FSWatcher, error) {
 	mw, err := fsnotify.NewWatcher()
 	if err != nil {
 		return nil, fmt.Errorf("failed to create new watcher: %v", err)
 	}
 
+	// activeWatchers is a map keyed by directory paths for directories that has been
+	// added to watch. The bool value is used to dedup multiple remove events for the same
+	// directory from fsnotify (inotify sends two events when removing a dir, "DELETE_SELF"
+	// and "DELETE,ISDIR" where the difference isn't exposed by fsnotify).
```
Comment (myitcv) [inline cmd/govim/internal/fswatcher/os_others.go:31]: Please can you explain in this comment _how_ the bool does that? i.e. what function does the use of a bool serve?

Comment (myitcv) [inline cmd/govim/internal/fswatcher/os_others.go:31]: Please can you explain in this comment _how_ the bool does that? i.e. what function does the use of a bool serve? That means you're not jumping below to find the reason :)

File: cmd/govim/internal/fswatcher/os_darwin.go [hunk 6]
```diff
@@ -13,17 +14,19 @@ import (
...
+)
 
 type fswatcher struct {
 	eventCh chan Event
 	es      *fsevents.EventStream
 }
 
-func New(gomodpath string, tomb *tomb.Tomb) (*FSWatcher, error) {
-	dirpath := filepath.Dir(gomodpath)
+func New(root string, filter watchFilterFn, logf logFn, tomb *tomb.Tomb) (*FSWatcher, error) {
```
Comment (myitcv) [inline cmd/govim/internal/fswatcher/os_darwin.go:28]: Can `root` be a filepath?

Would be good to get some docs for both `New()` functions on what `root` and `filter` do, and the values that `root` can take.

File: cmd/govim/internal/fswatcher/os_darwin.go [hunk 7]
```diff
@@ -13,17 +14,19 @@ import (
...
+	dirpath := filepath.Dir(root)
```
Comment (myitcv) [inline cmd/govim/internal/fswatcher/os_darwin.go:29]: We seem to calculate `filepath.Dir` here, but from what I can tell `root` will only ever be a directory? 

And why do we end up watching the parent directory of `root`? A comment here would be useful too.

File: cmd/govim/internal/fswatcher/os_others.go [hunk 8]
```diff
@@ -13,49 +16,157 @@ type fswatcher struct {
...
+	w := &FSWatcher{&fswatcher{
+		eventCh: eventCh,
+		errCh:   mw.Errors,
+		mw:      mw,
+		logf:    logf,
+	}}
+
+	var populateWatchers func(string) ([]string, error)
+	populateWatchers = func(initPath string) (files []string, err error) {
+		return files, filepath.Walk(initPath, func(path string, info os.FileInfo, err error) error {
```
Comment (myitcv) [inline cmd/govim/internal/fswatcher/os_others.go:44]: I'm not such a fan of `files` being populated as a side effect of the second return value here; it hurts readability to my mind. 

Can `populateWatchers` be a method on `fswatcher` - would that tidy things up a bit?

File: cmd/govim/watcher.go [hunk 9]
```diff
@@ -165,6 +87,59 @@ func (m *modWatcher) watch() {
...
+			if _, err := os.Stat(filepath.Join(root, curr, "go.mod")); err == nil {
+				return true
+			}
+		}
+		return false
+	}
+}
+
 func ofInterest(path string) bool {
 	return filepath.Ext(path) == ".go" || filepath.Base(path) == "go.mod" || filepath.Base(path) == "go.sum"
```
Comment (myitcv) [inline cmd/govim/watcher.go:144]: I'm not sure this is entirely true in the context of CGO. Indeed the client can't really know what is relevant or not. And with `//go:embed` we could need to care about any file, because removal of a file that is the source of a `//go:embed` will result in an error.

So this would point to us having to notify about _everything_. Perhaps we address that in a follow up PR, along with batching the notifications?

**Review by myitcv — APPROVED** [review]
Some small nits, and one more fundamental questions. But overall, I'm very supportive and grateful for you tidying up this code. It's one of those areas with loads of edge cases. So better to get this in sooner rather than later, and then iterate of further fixes guided by tests.

File: cmd/govim/internal/fswatcher/os_others.go (hunk 5 above)
Comment (leitzler) [reply cmd/govim/internal/fswatcher/os_others.go]: Duplicate comment :)

Comment (leitzler) [reply cmd/govim/internal/fswatcher/os_others.go]: Added an explanation.

File: cmd/govim/internal/fswatcher/os_others.go (hunk 8 above)
Comment (leitzler) [reply cmd/govim/internal/fswatcher/os_others.go]: Fixed as per our Slack conversation & moved `populatedWatchers` into `fswatcher`.

File: cmd/govim/watcher.go (hunk 9 above)
Comment (leitzler) [reply cmd/govim/watcher.go:144]: Interesting, this is unchanged code but that's a good point I didn't think about. Let's do it with the batching.

File: cmd/govim/internal/fswatcher/os_darwin.go (hunk 7 above)
Comment (leitzler) [reply cmd/govim/internal/fswatcher/os_darwin.go:29]: Good catch, it's a leftover from before when the go.mod path was passed in. We shouldn't watch the `root` parent, just the root. Fixed.

File: cmd/govim/internal/fswatcher/os_darwin.go (hunk 6 above)
Comment (leitzler) [reply cmd/govim/internal/fswatcher/os_darwin.go:28]: Added a description for both versions of `New` and added a sanity check that ensure that the root is a directory.

**Comment by myitcv** [comment]
Still LGTM 😄
